align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+            prompt_to_match - the prompt we are selecting ICL examples for
+        Output:
+            selected_icl_examples - relevant subset of self.icl_datasets
+        '''
+        # Check that we have incontext examples for the target kdma
+        if sys_kdma_name not in self.icl_datasets:
+            raise RuntimeError(f"No incontext samples for targeted kdma: {sys_kdma_name}")
+        n_icl_examples = self.incontext_settings["number"]
+        possible_icl_examples = self.icl_datasets[sys_kdma_name]
+        # Check that we have enough incontext examples for the target kdma
+        if len(possible_icl_examples) < n_icl_examples:
+            raise RuntimeError(f"Not enough possible incontext samples to learn from. Only "
+                            f"{len(possible_icl_examples)} samples available while asking for "
+                            f"{n_icl_examples} incontext samples.")
+        # If using LOO, don't include example ICL with exact same prompt
+        if self.incontext_settings.get("leave_one_out", False):
+            possible_icl_examples = [
+                icl_ex for icl_ex in possible_icl_examples
+                if icl_ex["prompt"] != prompt_to_match
+            ]
+
+        # Downselect to n_icl_examples via given method
+        icl_strategy = self.incontext_settings["method"]
+        
+        if icl_strategy == "random":
+            selected_icl_examples = random.sample(possible_icl_examples, n_icl_examples)
+        elif icl_strategy == "bert_similarity":
+            possible_icl_prompts = [icl_sample["prompt"] for icl_sample in possible_icl_examples]
+            # Create similarity scores between the ICL samples and find top-k indices
+            from bert_score import score
+            _, _, F1 = score([prompt_to_match]*len(possible_icl_prompts), possible_icl_prompts, lang="en")
+            _, indices = torch.topk(F1, n_icl_examples)
+
+            selected_icl_examples = [possible_icl_examples[i] for i in indices]
+        else:
+            raise ValueError(f'"{icl_strategy}" is not a valid incontext method. Please use "random" or '
+                                '"bert_similarity"')
+
+        return selected_icl_examples
+
+
+class ComparativeRegressionIncontextExampleGenerator(IncontextExampleGenerator):
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+    
+    def set_icl_datasets(self):
+        icl_datasets = {}
+        incontext_data = self.read_icl_dataset_files()
+        
+        # Add each target to icl_datasets
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            icl_datasets[sys_kdma_name] = []
+            kdma_incontext_data = incontext_data[sys_kdma_name]
```

Comment:
Might get a key error here. My earlier comment applies again

Reply:
I don't think we will get a key error because in  the `_read_icl_dataset_files()` function we do:
```
for target_kdma in self.target_kdmas:
    sys_kdma_name = target_kdma['kdma']
    incontext_data[sys_kdma_name] = []
```
There is a chance the list stays empty, but that will throw a runtime error in `select_icl_examples()`

Reply:
This comment is applicable if we load KDMAs based on incontext settings instead of target KDMAs